### DISCUSSION TITLE
Building Boost.Python for versions before 1.80.0

### DISCUSSION THREAD

**Discussion by devm18426** [MarkusJx/install-boost#38] [General]
I'm currently [trying](https://github.com/devm18426/stpyv8/blob/7b0ffc306e4eb66d14ddad7168350c4575f9c317/.github/workflows/build-windows.yml) and [failing miserably](https://github.com/devm18426/stpyv8/actions/runs/5883076461/job/15954949795) to build Boost.Python for Boost version 1.74 on Windows.

I was able to do it on my local machine (after MUCH trial and error) but on Github Actions it just won't go.

I'd really appreciate any tips on what I should be doing.

Thanks in advance.

**Comment by devm18426** [comment]
I think I got it working.

Initially I tried to download/install Boost with this action and then run `b2` for Boost.Python.

Unfortunately, it appears that the Boost archive provided by this Action doesn't have the Boost.Python sources in the top level `boost` directory:
![image](https://github.com/MarkusJx/install-boost/assets/125835885/1f7f29d3-4c5d-4658-9cc2-31a9a0b4b0a8)

From what I can tell, this causes `b2` to try symlinking those files from the `libs` directory using the Windows `mklink` command, which fails. I can't understand why.

When looking at the official Boost source directory structure (downloaded from [here](https://boostorg.jfrog.io/ui/repos/tree/General/main/release/1.74.0/source/boost_1_74_0.zip) for example), the `boost` directory contains everything (including the Boost.Python code):

![image](https://github.com/MarkusJx/install-boost/assets/125835885/a1762b4d-67b7-41d5-895d-9595b08c513b)

I don't know how this Action's archives were generated so I can't explain why the structures are different, or why `b2` causes issues.

So then I tried downloading the official ZIP and building from there and that did the trick.

**Reply by MarkusJx** [reply]
Just some clarifications:
* Boost is built inside the [`prebuilt-boost` repo](https://github.com/MarkusJx/prebuilt-boost) using [this workflow](https://github.com/MarkusJx/prebuilt-boost/blob/main/.github/workflows/build-boost.yml)
* The boost sources are from the [boost git repo](https://github.com/boostorg/boost) (this was the easiest way to get every version of boost)
* The build command for the static boost libraries is this one:
```
.\b2.exe install --build-dir='tmp' --prefix='.' variant='release,debug' address-model='32,64' link=${{matrix.LINK}} python='3.7,3.8,3.9,3.10,3.11' -j4 msvc stage
```

Python support was originally requested in https://github.com/MarkusJx/install-boost/discussions/25, but I've decided to only support python in boost versions 1.80.0 and higher in https://github.com/MarkusJx/install-boost/discussions/25#discussioncomment-4070328.
I'm not quite sure why I did this, probably because re-building everything is a huge pain and will break this action. I'm also pretty sure I tried building some earlier versions of boost with the new options and that failed for numerous reasons (can't recall why). That would also explain why there is a mingw version of boost for v1.74.0 but not for any other version below 1.78.0.

This is also documented in the readme
> Starting from boost version `1.80.0`, the pre-built binaries will be built with `boost.python`

And python support was reported working in https://github.com/MarkusJx/install-boost/discussions/25#discussioncomment-4075241 for boost v1.80.0. But I myself can't really say anything about this, as I don't know a thing about the boost python libs.

If you need python support for versions earlier than 1.80.0, I can try building an older version, but this may not work. I'd suggest trying to use version 1.80.0 or higher, if possible.

**Reply by devm18426** [reply]
Thanks for the clarification!

I would love to use one of the newer existing builds you provide, however I'm currently trying to compile a project that isn't mine. This project depends specifically on 1.74... So the only option I have currently is to build it myself. I really wanted to use your action for the build process but as stated, this raised issues. You have given me some material to investigate but the alternative solution I've found works well enough for now.

If you were to build Boost.Python for older versions (maybe even just 1.74 :) ) that would be ideal! But I appreciate it's a hassle, so I won't ask...

In the event that I revisit building Boost.Python with this Action in the future, I'll be sure to update.

**Reply by MarkusJx** [reply]
I've started a [build](https://github.com/MarkusJx/prebuilt-boost/actions/runs/5890037715/job/15974373687) for version 1.74.0, let's see how that'll play out. If the build doesn't fail, I'll probably just replace the windows binaries in order to prevent this action from breaking. This will lead to continuity errors but I'll deal with that probably never so it'll be fine.

I'll post an update once there is something to report

**Reply by MarkusJx** [reply]
Update: Didn't work, will try again later

**Reply by devm18426** [reply]
Markus, you're amazing. Thanks for the effort! Much appreciated. Hope you're able to work it out.

**Reply by MarkusJx** [reply]
I've uploaded a version of boost 1.74.0 for windows server 2019 and support for python 3.7, 3.8 and 3.9. Python 3.10 and up unfortunately don't work.

**Reply by devm18426** [reply]
That's great Markus, thanks so much.

Regarding 3.10, if you're interested, it's possible to very easily patch the code so it compiles. The issue stems from usage of a function in the Python API which became private in 3.10: `_Py_fopen`. In [this](https://github.com/boostorg/python/commit/cbd2d9f033c61d29d0a1df14951f4ec91e7d05cd) change you may see how they fixed this issue in Boost.Python 1.76. Essentially changing the use of `_Py_fopen` with a standard `fopen` in a specific file, `exec.cpp`.

I can confirm that making this small change in 1.74 works just as well. You may see how easily I achieved this in Powershell with a simple find/replace prior to building:

    (Get-Content libs\python\src\exec.cpp).replace('_Py_fopen', 'fopen') | Set-Content libs\python\src\exec.cpp

https://github.com/devm18426/stpyv8/actions/runs/5904217668/workflow#L115

That said, I haven't tried this trick with 3.11.

Thanks again
